atomicapp/index.py

```diff
@@ -0,0 +1,204 @@
+        properly formatted way.
+        """
+
+        # In order to "format" it correctly, find the largest length of 'name', 'id', and 'appversion'
+        # Set a minimum length of '7' due to the length of each column name
+        id_length = 7
+        app_length = 7
+        location_length = 7
+
+        # Loop through each 'nulecule' and retrieve the largest string length
+        for entry in self.index["nulecules"]:
+            id = entry.get('id') or ""
+            version = entry['metadata'].get('appversion') or ""
+            location = entry['metadata'].get('location') or INDEX_DEFAULT_IMAGE_LOCATION
+
+            if len(id) > id_length:
+                id_length = len(id)
+            if len(version) > app_length:
+                app_length = len(version)
+            if len(location) > location_length:
+                location_length = len(location)
+
+        # Print out the "index bar" with the lengths
+        index_format = ("{0:%s}  {1:%s}  {2:10} {3:%s}" % (id_length, app_length, location_length))
+        print(index_format.format("ID", "VER", "PROVIDERS", "LOCATION"))
+
+        # Loop through each entry of the index and spit out the formatted line
+        for entry in self.index["nulecules"]:
+            # Get the list of providers (first letter)
+            providers = ""
+            for provider in entry["providers"]:
+                providers = "%s,%s" % (providers, provider[0].capitalize())
+
+            # Remove the first element, add brackets
+            providers = "{%s}" % providers[1:]
+
+            # Retrieve the entry information
+            id = entry.get('id') or ""
+            version = entry['metadata'].get('appversion') or ""
+            location = entry['metadata'].get('location') or INDEX_DEFAULT_IMAGE_LOCATION
+
+            # Print out the row
+            print(index_format.format(
+                id,
+                version,
+                providers,
+                location))
+
+    def update(self, index_image=INDEX_IMAGE):
+        """
+        Fetch the latest index image and update the file based upon
+        the INDEX_IMAGE attribute. By default, this should pull the
+        'official' Nulecule index.
+        """
+
+        logger.info("Updating the index list")
+        logger.info("Pulling latest index image...")
+        self._fetch()
+        logger.info("Index updated")
+
+    def generate(self, location, output_location=INDEX_GEN_DEFAULT_OUTPUT_LOC):
+        """
+        Generate an index.yaml with either a provided Git URL or path location
+        """
+        logger.info("Generating index.yaml from %s" % location)
+        self.index = deepcopy(self.index_template)
+        if not os.path.exists(location):
+            location = self._get_repo(location)
+
+        if not os.path.isdir(location):
+            raise Exception("Location must be a git repository")
+
+        for f in os.listdir(location):
+            nulecule_dir = os.path.join(location, f)
```

Comment:
Yes. We assume each directory has a Nulecule component for the generation command.

From your comment down below (in regards to the Git dependency) we can instead create a generation script inside of the projectatomic/nulecule-library repo instead and further improve it.